app/unit-tests/src/org/commcare/android/tests/formsave/CyclicCasesPurgeTest.java

```diff
@@ -0,0 +1,51 @@
+package org.commcare.android.tests.formsave;
+
+import android.view.View;
+import android.widget.TextView;
+
+import org.commcare.CommCareApplication;
+import org.commcare.CommCareTestApplication;
+import org.commcare.activities.FormEntryActivity;
+import org.commcare.android.util.ActivityLaunchUtils;
+import org.commcare.android.util.TestAppInstaller;
+import org.commcare.android.util.TestUtils;
+import org.commcare.dalvik.R;
+import org.commcare.models.AndroidSessionWrapper;
+import org.commcare.session.CommCareSession;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.annotation.Config;
+
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+/**
+ * Test Scenarios related to cyclic case relationships encountered during the case purge process.
+ * These tests are only relevant when the case purge process is enabled with property `cc-auto-purge`
+ */
+@Config(application = CommCareTestApplication.class)
+@RunWith(AndroidJUnit4.class)
+public class CyclicCasesPurgeTest {
+
+    @Before
+    public void setup() {
+        TestAppInstaller.installAppAndLogin(
+                "jr://resource/commcare-apps/cyclic_case_purge/profile.ccpr",
+                "test", "123");
+    }
+
+    @Test
+    public void testFormSaveResultingIntoCaseCycles_ShouldFail() {
+        TestUtils.processResourceTransactionIntoAppDb("/inputs/case_create_for_cyclic_case_purge.xml");
+        AndroidSessionWrapper sessionWrapper =
+                CommCareApplication.instance().getCurrentSessionWrapper();
+        CommCareSession session = sessionWrapper.getSession();
+        session.setDatum("case_id", "fe9adf5c-7661-4f8e-a48e-4059fcf1f9dd");
+        session.setCommand("m2-f0");
+        FormEntryActivity formEntryActivity = ActivityLaunchUtils.launchFormEntry();
+        View finishButton = formEntryActivity.findViewById(R.id.nav_btn_finish);
+        finishButton.performClick();
+        String message = ((TextView)formEntryActivity.getCurrentAlertDialog().getUnderlyingDialog().getDialog().findViewById(R.id.dialog_message)).getText().toString();
+        assert message.contentEquals(formEntryActivity.getString(R.string.invalid_case_graph_error));
```

Comment:
Out of curiosity, what is the *resulting* state we expect after this point?

It's good if this creates an error, but we should also probably confirm that after the error, a subsequent purge that is run directly succeeds, to ensure that the error prevented a bad state from being create.

Reply:
Thanks [added](https://github.com/dimagi/commcare-android/pull/2504/commits/dab3d9f007b0af9089d6e2c1d7e781d8529a880d) couple checks to verify the same.

An added point - purge process today only gets triggered on a form save when a current index is strictly [replaced](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/xml/CaseXmlParser.java#L281) and not if we simply add an index. This is why I have a dummy index [here](https://github.com/dimagi/commcare-android/blob/formplayerMerge/app/unit-tests/resources/inputs/case_create_for_cyclic_case_purge.xml#L31) in the restore file.  This would result into most form save causing a cycle to escape without errors even when the case purge is turned on so want to make sure that it's something intentional ?

Reply:
Awesome, these look good.

I believe it's not possible in the current algorithms for a new index to create a situation where cases need to be removed, which was the only old function of the purge call, but you're very correct that that does prevent this check from occurring.

If it's possible and easy, I think it's fine to add "creating a new case index" as one of the conditions which triggers a purge as the best incremental change.

Reply:
Got it, made the change in https://github.com/dimagi/commcare-core/pull/1004/commits/a148f7253e40f4e179998008b4ca05a26a19079a